Update to ESR 115

ptomato opened the pull request:
Credit to @xry111 in #71, which I had to close to retarget to the `next` branch.

I added two more sections to the migration guide, so it's worth another review.

xry111 committed:
Document the API change on JS_InitClass

xry111 committed:
Update to ESR 115

xry111 committed:
examples: Adapt JS_InitClass for ESR 115

xry111 committed:
Add mozilla::Tuple -> std::tuple into Migration Guide for 115

ptomato committed:
Add JSClass check to resolve example

This was a pre-existing problem in the resolve.cpp example. Before calling
getPriv(), we should check that the this-object has the correct JSClass.

ptomato committed:
Don't store value in prototype reserved slot

The prototype is a plain object now, and doesn't have a reserved slot.

ptomato committed:
Document change in bound function objects

Noticed in https://github.com/Distributive-Network/PythonMonkey/pull/74

arai-a commented on docs/Building SpiderMonkey.md:
```diff
@@ -5,14 +5,10 @@ Use these instructions to build your own copy of SpiderMonkey.
 
 You will need a **C++ compiler** that can handle the C++17 standard, **Rust**
-version 1.59 or later, **GNU Make**, **zlib**, and **libffi**.
+version 1.66 or later, **GNU Make**, **zlib**, and **libffi**.
```
how about adding a link to https://searchfox.org/mozilla-esr115/rev/61b47de1faebf23626e519b2464b461589fbea3e/python/mozboot/mozboot/util.py#14 to clarify where it's required?

arai-a commented on docs/Building SpiderMonkey.md:
```diff
@@ -5,14 +5,10 @@ Use these instructions to build your own copy of SpiderMonkey.
-than just `autoconf`.
 
-> **NOTE** SpiderMonkey also requires ICU of at least version 71.1, but
+> **NOTE** SpiderMonkey also requires ICU of at least version 73.1, but
```
here too, how about adding a link to https://searchfox.org/mozilla-esr115/rev/61b47de1faebf23626e519b2464b461589fbea3e/js/moz.configure#1107 ?

arai-a commented on docs/Building SpiderMonkey.md:
```diff
@@ -22,18 +18,17 @@ than just `autoconf`.
-   defined](https://bugzilla.mozilla.org/show_bug.cgi?id=1780857)
+At the time of writing, the latest source for Firefox ESR 115, which
+contains the source for SpiderMonkey ESR 115, can be found here:
+https://ftp.mozilla.org/pub/firefox/releases/115.0esr/source/
```
this could be updated to https://ftp.mozilla.org/pub/firefox/releases/115.1.0esr/source/ ?

arai-a commented on docs/Building SpiderMonkey.md:
```diff
@@ -22,18 +18,17 @@ than just `autoconf`.
+the fixes for the following bug:
+
+ - [`js_new<JSRuntime>` allocates memory block not satisfying
+   `alignof(JSRuntime)`](https://bugzilla.mozilla.org/show_bug.cgi?id=1841040)
```
Once the link above is updated to 115.1.0esr, this section can be removed, given the patch is uplifted to esr115 and released as 115.1.0.

or maybe this section could be kept with a comment this this applies only to 115.0.x

ptomato replied:
I think it's best to keep it uncomplicated if it's not necessary with the latest version, so I removed this section.

arai-a commented on docs/Migration Guide.md:
```diff
@@ -5,11 +5,78 @@ of SpiderMonkey to the next ESR version.
+the same `JSClass` and `JSClassOps` specification.
+If the prototype should behave differently than the instances, custom
+checks were needed.
+Now `JS_InitClass` accepts a `const char *` parameter for naming the
+constructor function, so a different `JSClass *` can be used for the
+prototype object.
+If `nullptr` is passed, the prototype object will be a plain JS object.
```
how about adding a link https://searchfox.org/mozilla-esr115/rev/61b47de1faebf23626e519b2464b461589fbea3e/js/src/jsapi.h#322-344 ?

ptomato replied:
I added a link to the patch, because I think that makes it clearer what exactly changed.

arai-a commented on docs/Migration Guide.md:
```diff
@@ -5,11 +5,78 @@ of SpiderMonkey to the next ESR version.
 This is a non-exhaustive list of minor API changes and renames.
 
 - `JS::ModuleInstantiate` → `JS::ModuleLink`
+- `mozilla::Tuple` → `std::tuple` (`<mozilla/Tuple.h>` is removed, use
+  the standard C++ header `<tuple>` instead)
```
also `mozilla::IsFinite` and `mozilla::IsNaN` are removed, in favor of `std::isfinite` and `std::isnan`.

arai-a replied:
Also, it would be nice to mention that the filename used by the engine is UTF-8 ([bug 1492090](https://bugzilla.mozilla.org/show_bug.cgi?id=1492090))

such as `JS::CompileOptions`.

https://searchfox.org/mozilla-esr115/rev/61b47de1faebf23626e519b2464b461589fbea3e/js/public/CompileOptions.h#120,126-130
```cpp
class JS_PUBLIC_API TransitiveCompileOptions {
...
  // UTF-8 encoded file name.
  const char* filename_ = nullptr;

  // UTF-8 encoded introducer file name.
  const char* introducerFilename_ = nullptr;
```

arai-a commented on docs/Migration Guide.md:
```diff
@@ -5,11 +5,78 @@ of SpiderMonkey to the next ESR version.
+
+### Module Import ###
+
+Previously, `JS::FinishDynamicModuleImport()` when passed an
+`evaluationPromise` argument that was null, would take the pending
+exception from the context and reject the import promise with it.
+This is no longer the case.
```
I haven't found the underlying change for this behavior change.
I'll look into this tomorrow.

ptomato replied:
I actually cannot reproduce it with a minimal example. So maybe I misunderstood the behaviour change. In case it helps figure out what was going on, here's the change I had to make: https://gitlab.gnome.org/GNOME/gjs/-/commit/572073dff01c1cf199c08a4e537f13ad5f99a290

Maybe I was just originally misusing the API :shrug: If we can't figure it out, I can drop this section from the migration guide.

arai-a replied:
tested with mozjs 115.1.0, and the test added by the patch passes even without the patch for `gjs/module.cpp` file applied.
is there more context about what went wrong with unpatched code?

ptomato replied:
Hmm! At the time, I was getting an exception without a message, instead of the exception thrown by the module import, so the test would fail. I wonder if it was just a regression in the 115.0 alpha which I was testing with. Anyway, it sounds like I should drop this section. Thanks for looking into it!

arai-a reviewed (commented):
Thank you!
Mostly looks good.
I'll check the module behavior and finish the review tomorrow.

ptomato committed:
Link to 115.1.0esr source

This version no longer requires applying the patch from bug 1841040.

ptomato committed:
Linkify minimum versions of dependencies

This adds links to the places in the Mozilla source code where the
minimum versions of Rust and ICU are defined.

ptomato committed:
Document more minor API changes in migration guide

ptomato commented:
Thanks for the review!
